### ISSUE TITLE
"Sync Sound" + "Automatic, skip at most" does the same thing as force frame skip

### ISSUE THREAD

**Issue by SkyPirate1** [snes9xgit/snes9x#756]
If I set the auto frameskip to anything other than zero while the Sync Sound option is enabled, frame skipping is forced to whatever that value is.

So if it is set to 1, I get 60 fps 30/60 or 50 fps 25/50 for PAL.
If set to 2, I get 60 fps 20/60 etc.

Now if I pause emulation and unpause, it goes back up to what it should be for a little bit and then drops again, or if I reset the game it drops again.  In TMNT Turtles in Time for example, the framerate drops when it goes back to the Konami logo screen while in demo mode.

Now I know the config file comment says something about if auto frame skip is set too low then the FPS value on screen will be inaccurate, but if I set auto skip to 10 you can see plain as day it really is dropping 10 frames.  This happens regardless if fullscreen, emulated fullscreen or if using a multiplier (1x, 2x, etc).  OpenGL or D3D output also makes no difference.  I'm running Windows 8.1 64 bit.

I can get smooth 60 fps in this emulator otherwise if it is not in true fullscreen, so what gives?

**Comment by bearoso** [comment]
Sync sound interacts badly with other framerate limiters. In the Gtk port it's just another type of frame limiter separate from automatic and fixed, for example. If you enable sync sound, you should run with Fixed frame skip 0 only with vsync enabled. That's the only scenario where it makes any sense.

**Comment by SkyPirate1** [comment]
I usually only run it with fixed rate set to 0 and no vsync, but over the years there's always been a slight stutter to the framerate (similar to the problem in mGBA) and I always notice it on the Konami logo when the bar rolls down the screen at the beginning of TMNT IV.  So I tried to fiddle with the settings to try to get it smoother, and vsync works for that, but with all the talk about "input lag" I leave vsync off just in case, and I turn on "Reduce Input Lag".  If this is something that isn't really fixable then perhaps there should be a note on the auto frameskip option in the gui that it is not recommended in conjunction with sync sound? Or maybe auto disable one when the other is used?

And with that I want to ask some questions.  If the SNES and NES have an output of 60.09881Hz, and old CRT televisions rendered at 59.94Hz, how did that pan out?  How did the televisions deal with that extra frame?  Or is it that frequency is not exactly indicative of framerate?

**Comment by bearoso** [comment]
Yes, many emulators now account for this. See https://github.com/snes9xgit/snes9x/wiki/Audio-is-crackling-or-skipping. The Automatic Frame Rate mode *is* a drop-frame mode. It drops a frame when it gets too far behind. If you have vsync off, however, there's no need, because it doesn't have to wait to display the frame. Same with reduce input lag--all the setting does is adjust where in time the frame is processing after vsync so it can have more up-to-date controller data when it appears out on the screen. If you don't use vsync it just goes on ahead. I suggest turning on vsync to get rid of the tearing and microstutter. With the reduce input lag setting on it's not much additional latency.

The double-strike mode of the SNES was nearly perfectly compatible. TVs would usually adjust. In fact, your monitor can probably do the same. On Windows with ToastyX's CRU you can create a video mode that has the right frequency, or on X11 you can just create a modeline. I do it all the time. Just copy your desktop mode and tweak it slightly. You can usually adjust the pixel clock in intervals of about 10khz, so it can be pretty precise just that way, or you can work with the sync intervals a bit and get *really* close to the right rate. Otherwise, turning on the automatic input rate in sound settings will detect your monitor's actual frequency and slow down emulation and stretch the audio to fill the gaps.

**Comment by bearoso** [comment]
BTW, there was a problem with the double-strike mode that cropped up back then, but it was on the Playstation. The mode used had too small of a vertical blanking interval, and the RCA/Zenith controller couldn't handle it. So the image would bounce up and down the screen.

**Comment by OV2** [comment]
As bearoso said, combining sync sound with a skip > 0 does not make much sense, and as you determined does not work in a useful way. Having it at auto with max skipped frames set to 0 makes sense in that it enables the speed up/down feature to work correctly.

It does not work with a number > 0 because it uses a simple timer to determine if the last frame took longer to emulate than what it expects, and does not account for waiting on the sound. So every time it has to wait for sound samples to be consumed it will drop a frame.

**Comment by SkyPirate1** [comment]
Thanks for the explanations gentlemen ;)  And thanks for keeping snes9x going all these years.  I was getting worried it was starting to fade away especially with the long time since the previous release.  I never liked bsnes/higan/ares gui but I do appreciate what they have done for accuracy.
